core/common-util/src/main/java/uk/gov/gchq/gaffer/commonutil/collection/LimitedSortedList.java

```diff
@@ -0,0 +1,300 @@
+    private final Integer limit;
+    private LinkedList<E> backingList;
+
+    // Used to deduplicate elements if required
+    private HashMap<E, Object> backingHashSet;
+
+    public LimitedSortedList(final Comparator<E> comparator) {
+        this(comparator, null);
+    }
+
+    public LimitedSortedList(final Comparator<E> comparator, final Integer limit) {
+        this(comparator, limit, false);
+    }
+
+    public LimitedSortedList(final Comparator<E> comparator, final Integer limit, final boolean deduplicate) {
+        this(comparator, limit, deduplicate, new LinkedList<>());
+    }
+
+    public LimitedSortedList(final Comparator<E> comparator, final Integer limit, final boolean deduplicate, final LinkedList<E> backingList) {
+        if (null == comparator) {
+            throw new IllegalArgumentException("Comparator is required");
+        }
+        if (null != limit && limit < 1) {
+            throw new IllegalArgumentException("Limit should be more than 0");
+        }
+        if (null == backingList) {
+            throw new IllegalArgumentException("backingList is required");
+        }
+        this.comparator = comparator;
+        this.limit = limit;
+        this.deduplicate = deduplicate;
+        this.backingList = backingList;
+
+        if (deduplicate) {
+            backingHashSet = new HashMap<>();
+        }
+    }
+
+    public boolean add(final E e) {
+        if (deduplicate && backingHashSet.containsKey(e)) {
+            return false;
+        }
+
+        if (null == limit || backingList.size() < limit) {
+            orderedAdd(e);
+            return true;
+        } else {
+            if (comparator.compare(backingList.peekLast(), e) > 0) {
+                final E removedItem = backingList.removeLast();
+                if (deduplicate) {
+                    backingHashSet.remove(removedItem);
+                }
+                orderedAdd(e);
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    private void orderedAdd(final E e) {
+        int index = 0;
+        for (final E existingE : backingList) {
+            if (comparator.compare(existingE, e) >= 0) {
+                break;
+            }
+            index++;
+        }
+        backingList.add(index, e);
+        if (deduplicate) {
+            backingHashSet.put(e, PRESENT);
+        }
+    }
+
+    @Override
+    public boolean remove(final Object o) {
+        if (deduplicate) {
+            backingHashSet.remove(o);
+        }
+        return backingList.remove(o);
+    }
+
+    @Override
+    public boolean containsAll(final Collection<?> c) {
+        if (deduplicate) {
+            backingHashSet.keySet().containsAll(c);
+        }
+        return backingList.containsAll(c);
+    }
+
+    @Override
+    public boolean addAll(final Collection<? extends E> c) {
+        boolean result = false;
+        for (final E e : c) {
+            if (add(e)) {
+                result = true;
```

Comment:
Do we want to keep a track of the result like this in the removeAll method too? (and any other method which is expected to report whether an operation on the collection was successful)